### PR TITLE
Add Contains, StartsWith and EndsWith

### PR DESCRIPTION
**Pull request by netdragoon**
https://github.com/Flepper/flepper/issues/57

### CODE REVIEW

**Comment by netdragoon** [comment]
@AlbertoMonteiro Do you have any problems with the code?

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 1]
```diff
@@ -238,5 +238,64 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+        [Fact]
+        public void ShoulContainLikeContains()
+        {
+            FlepperQueryBuilder.Select()
+                .From("table")
+                .Where("field").Contains("abc")
+                .Build()
+                .Trim()
+                .Should()
+                .Contain("LIKE");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Test the full SQL Query and also check parameters

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 2]
```diff
@@ -238,5 +238,64 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+        [Fact]
+        public void ShoulContainLikeStartsWith()
+        {
+            FlepperQueryBuilder.Select()
+                .From("table")
+                .Where("field").StartsWith("abc")
+                .Build()
+                .Trim()
+                .Should()
+                .Contain("LIKE");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Test the full SQL Query and also check parameters

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 3]
```diff
@@ -238,5 +238,64 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+        [Fact]
+        public void ShoulContainLikeEndsWith()
+        {
+            FlepperQueryBuilder.Select()
+                .From("table")
+                .Where("field").EndsWith("abc")
+                .Build()
+                .Trim()
+                .Should()
+                .Contain("LIKE");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Test the full SQL Query and also check parameters

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 4]
```diff
@@ -238,5 +238,64 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+        public void ShoulContainLikeContainsAndStartsWithAndEndsWith()
+        {
+            var result = FlepperQueryBuilder.Select()
+                .From("table")
+                .Where("field1").Contains("abc")
+                .And("field2").StartsWith("abc")
+                .And("field3").EndsWith("abc")
+                .Build()
+                .Trim()
+                .Should();
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Test the full SQL Query and also check parameters

**Comment by AlbertoMonteiro** [comment]
@netdragoon you just need to write unit tests properly, the implementation is ok.

**Comment by netdragoon** [comment]
AlbertoMonteiro how can I exclude this PR?

**Comment by netdragoon** [comment]
One remark has a problem in the autal code, in two moments two codes are the same, this could be revised.

@AlbertoMonteiro

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 1 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 2 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 3 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 4 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 1 above)
Comment (AlbertoMonteiro) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 2 above)
Comment (AlbertoMonteiro) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 3 above)
Comment (AlbertoMonteiro) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 5]
```diff
@@ -242,60 +242,87 @@ public void ShoulContainWhereWithNotEqualToNull()
...
-                .And("field2").StartsWith("abc")
-                .And("field3").EndsWith("abc")
-                .Build()
+                .Where("field1").Contains("abc1")
+                .And("field2").StartsWith("abc2")
+                .And("field3").EndsWith("abc3")
+                .BuildWithParameters();               
+
+            result.Query.Should()
+                .Contain("@p0");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 6]
```diff
@@ -242,60 +242,87 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+            result.Query.Should()
+                .Contain("@p1");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 7]
```diff
@@ -242,60 +242,87 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+            result.Query.Should()
+                .Contain("@p2");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 8]
```diff
@@ -242,60 +242,87 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+
+            result.Query.Should()
+                .Contain("[field1]");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 9]
```diff
@@ -242,60 +242,87 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+            result.Query.Should()
+                .Contain("[field2]");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 10]
```diff
@@ -242,60 +242,87 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+            result.Query.Should()
+                .Contain("[field3]");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: You still must test the FULL SQL Query, do not use Contains

**Comment by AlbertoMonteiro** [comment]
@netdragoon you also must increase coverage.

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 5 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok

Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: I did not understand

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 6 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: I did not understand

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 7 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: I did not understand

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 8 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: I did not understand

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 9 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: I did not understand

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 10 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: I did not understand

**Comment by netdragoon** [comment]
I did not understand (you also must increase coverage.)?????

@AlbertoMonteiro

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 5 above)
Comment (gustavoferrazfontes) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: We would like that you test the full sql query generated by fluent builder.

see example bellow
```
                  QueryResult result = FlepperQueryBuilder
                 .Select()
                 .From("table")
                 .Where("field1").Contains("abc1")
                 .And("field2").StartsWith("abc2")
                 .And("field3").EndsWith("abc3")
                 .BuildWithParameters(); 
                 
               result
               .Trim()
               .Should()
              .Be("SELECT * [table] WHERE field1 LIKE @p1 AND field2 LIKE @p2 AND field3 LIKE @p3")
```

Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: @gustavoferrazfontes Ok, Test done

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 6 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok, Test done

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 7 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok, Test done

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 8 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok, Test done

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 9 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok, Test done

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 10 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok, Test done

**Comment by netdragoon** [comment]
@AlbertoMonteiro and @gustavoferrazfontes 


Which would be for the method below the satisfactory test, can you cite an example?

```
/// <summary>
/// 
/// </summary>
/// <typeparam name="T"></typeparam>
/// <param name="whereFilter"></param>
/// <param name="value"></param>
/// <returns></returns>
public static IComparisonOperators Contains<T>(this IWhereFilter whereFilter, T value)
	=> whereFilter.To<ComparisonOperators>().Contains(value);
```
???

**Comment by netdragoon** [comment]
@AlbertoMonteiro e @gustavoferrazfontes 

If for some reason, do not want this PR and if they can say it, I remove!

 ???

**Comment by gustavoferrazfontes** [comment]
@netdragoon 
What we  are trying to say is just for you test de full sql statement generated by fluent builder. 
Because in the first PR that you sended to us, you just tested a piece of the query.

We don't have any problem with your contributor, just comments 😄

**Comment by netdragoon** [comment]
Ok @gustavoferrazfontes

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 11]
```diff
@@ -339,5 +339,67 @@ public void ShoulContainLikeContainsAndStartsWithAndEndsWith()
...
+            result.Query.Trim()
+                .Contains("BETWEEN");
+            result.Query.Trim()
+                .Contains("@p0");
+            result.Query.Trim()
+                .Contains("@p1");
+            result.Query.Trim()
+                .Contains("[table]");
+            result.Query.Trim()
+                .Contains("[field]");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Remove those lines, 351...360, the 365 is enough!

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 12]
```diff
@@ -339,5 +339,67 @@ public void ShoulContainLikeContainsAndStartsWithAndEndsWith()
...
+            result.Query.Trim()
+                .Contains("@p0");
+            result.Query.Trim()
+                .Contains("@p1");
+            result.Query.Trim()
+                .Contains("@p2");
+            result.Query.Trim()
+                .Contains("[table]");
+            result.Query.Trim()
+                .Contains("[field]");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Remove those lines, 381...392, the 397 is enough!

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs [hunk 13]
```diff
@@ -238,5 +238,168 @@ public void ShoulContainWhereWithNotEqualToNull()
...
+        {
+            QueryResult result = FlepperQueryBuilder.Select()
+                .From("table")
+                .Where("field").Contains("abc")
+                .BuildWithParameters();
+
+            result.Query
+                .Trim()
+                .Should()
+                .Contain("LIKE");
```
Comment (AlbertoMonteiro) [inline Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Remove those lines, 250...253, the 258 is enough!
Repeat this same logic to other tests.

Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 12 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok

File: Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs (hunk 11 above)
Comment (netdragoon) [reply Flepper.Tests.Unit/QueryBuilder/Commands/WhereFilterTests.cs]: Ok

**Comment by AlbertoMonteiro** [comment]
Nice @netdragoon thats all fine now! Thanks for your contribution

**Pull request merged**